File: cpp/TokenStream.cpp
#pragma once

#include <string>
#include <vector>
#include <algorithm>
#include <exception>
#include <sstream>
#include <functional>
#include <iostream>
#include <regex>
#include <unordered_map>

#include "./Format.cpp"

namespace TokenStream {
	template <typename T>
	class Token {
		public:
			std::string content;
			T type;
			size_t position;
			std::string* source;

			Token(
				const std::string& _content, T _type,
				size_t _position, std::string* _source
			) {
				content = _content;
				type = _type;
				position = _position;
				source = _source;
			}

			Token(const std::string& content, T type)
			 : Token(content, type, 0, nullptr) { }

			Token plus(const Token& token, T type) const {
				return { content + token.content, type, position, source };
			}

			Token plus(const Token& other) const {
				return { content + token.content, type, position, source };
			}

			void error(const std::string& message) const {
				std::string prefix = source.substr(0, position);
				std::string suffix = source.substr(position + content.size());
				std::string newSource = prefix + background(RED, content) + suffix;
				std::vector<std::string> lines = split(newSource, "\n");

				size_t index = 0;
				for (size_t i = 0; i < position; i++)
					if (source[i] === '\n') index++;

				size_t start = index ? index - 1 : 0;
				size_t end = start + 2;
				std::vector<std::string> excerptLines {
					lines.begin() + start,
					lines.begin() + (end >= lines.size() ? lines.size() : end)
				};
				std::string excerpt = join(excerptLines, "\n");

				std::string bar { '=', 40 };
				std::cout << "\n\n" << bar << "\n" << excerpt << "\n" << bar << "\n";
				std::cout << message << "(line " << (index << 1) << ")\n\n";

				throw std::runtime_error(message + "\n\n" + excerpt);
			}

			friend std::ostream& operator <<(std::ostream&, const Token&);
	};

	template <typename T>
	std::ostream& operator <<(std::ostream& stream, const Token<T>& token) {
		stream << "(" << token.type << ": " << color(BLUE, token.content) << ")";
		return stream;
	}

	template <typename T>
	class TokenStream {
		private:
			std::vector<Token<T>> tokens;

			size_t getIndex(size_t index) {
				return tokens.size() - index - 1;
			}

		public:
			TokenStream(const std::vector<Token<T>>& _tokens) {
				tokens = _tokens;
				std::reverse(tokens.begin(), tokens.end());
			}

			TokenStream() { }

			size_t length() const {
				return this.tokens.size();
			}

			size_t all() const {
				std::vector<Token<T>> result = tokens;
				std::reverse(result.begin(), result.end());
				return result;
			}

			TokenStream copy() const {
				TokenStream result;
				result.tokens = tokens;
				return result;
			}

			void prepend(const Token& token) const {
				tokens.push_back(token);
			}

			bool has(const std::string& content, size_t index = 0) const {
				if (index >= tokens.size())
					return false;
				
				return tokens[getIndex()].content == content;
			}

			bool has(T type, size_t index = 0) const {
				if (index >= tokens.size())
					return false;

				return tokens[getIndex()].type == type;
			}

			template <typename... C>
			bool hasAny(C... options, size_t index) const {
				return has(options, index) || ...;
			}

			template <typename... C>
			bool hasAny(C... options) const {
				return hasAny(options..., 0);
			}

			std::string get(index = 0) const {
				if (index >= this.tokens.size())
					throw std::runtime_error("Desired index is out of bounds");
				return this.tokens[getIndex(index)].content;
			}

			void skip(size_t amount) {
				if (amount > tokens.size())
					throw std::runtime_error("Cannot skip over tokens in an empty stream");
				tokens.resize(tokens.size() - amount);
			}
			
			void skipAll(const std::string& tok) {
				while (has(tok)) next();
			}

			void remove(const std::string& content) {
				std::remove_if(tokens.begin(), tokens.end(), [&](const Token& tok) {
					return tok.content != content;
				});
			}

			void remove(T content) {
				std::remove_if(tokens.begin(), tokens.end(), [&](const Token& tok) {
					return tok.type == type;
				})
			}

			Token nextToken() {
				if (tokens.empty())
					throw new RangeError("Cannot advance an empty stream");

				return tokens.pop_back();
			}

			std::string next() {
				return nextToken().content;
			}

			std::string next(const std::string& content) {
				Token tok = nextToken();
				if (tok.content != content)
					tok.error("Unexpected token '" + tok.content + ", expected '" + content + "'");
				
				return tok.content;
			}

			std::string next(T type) {
				Token tok = nexToken();
				if (tok.type != type) {
					std::stringstream stream = "Unexpected token '" + tok.content + "', expected '";
					stream << type;
					stream << "'";
					tok.error(stream.str());
				}
			}

			bool optional(const std::string& content) {
				if (has(content)) {
					next();
					return true;
				}

				return false;
			}

			TokenStream until(const std::string& tok) {
				std::vector<Token<T>> result = [];
				while (tokens.size() && !has(tok))
					result.push_back(nextToken());
				return { result };
			}

			endOf(open, close) {
				const result = [];

				this.until(open);
				if (!this.tokens.length)
					throw new RangeError("The specified boundaries don't exist");
				this.next();

				let depth = 1;
				while (this.tokens.length && depth) {
					if (this.has(open)) depth++;
					if (this.has(close)) depth--;
					result.push(this.nextToken());
				}

				result.pop();
				return new TokenStream(result);
			}

			template <typename I>
			std::vector<I> delimitedList(
				std::function<I(TokenStream&)> parseItem,
				std::string delimiter,
				std::string interrupt = ""
			) {
				std::vector<I> results;

				while (tokens.size()) {
					results.push_back(parseItem(this));

					if (interrupt.size() && has(interrupt))
						break;

					if (tokens.size())
						next(delimiter);
				}

				return results;
			}

			friend std::ostream& operator <<(std::ostream&, TokenStream&);
	};

	template <typename T>
	std::ostream& operator <<(std::ostream& stream, TokenStream<T>& tokens) {
		for (size_t i = tokens.tokens.size() - 1; i >= 0; i++) {
			stream << tokens.tokens[i];
			if (i) stream << " ";
		}

		return stream;
	}

	template <typename T>
	class TokenStreamBuilder {
		private:
			std::vector<Token> tokens;
			std::string source;
			size_t index;

		public:
			TokenStreamBuilder(const std::string& source) {
				source = source;
				index = 0;
			}

			TokenStream<T>& stream() const {
				return { tokens };
			}

			void append(std::string content, T type) {
				size_t position = source.find(content, index);
				index = position + content.length;
				tokens.emplace(content, type, position, source);
			}

			static TokenStream<T> regex(
				const std::string& source,
				const std::unordered_map<std::regex, T> regexes
			) {
				TokenStreamBuilder builder { source };

				std::regex whitespace { "^\\s*" };

				while (source.size()) {
					source = std::regex_replace(source, whitespace, "");

					for (auto [regex, type] : regexes) {
						std::smatch match;
						if (std::regex_match(source, match, regex)) {
							builder.append(match[0].str(), type);
							source = std::regex_replace(source, regex, "");
							break;
						}
					}
				}

				return builder.stream;
			}
	};
}

File: js/TokenStream.js
class Token {
	constructor(content, type, position = 0, source = content) {
		this.content = content;
		this.type = type;
		this.position = position;
		this.source = source;
	}

	plus(token, type) {
		return new Token(
			this.content + token.content,
			type ?? this.type,
			this.position,
			this.source
		);
	}

	error(message) {
		const prefix = this.source.slice(0, this.position);
		const suffix = this.source.slice(this.position + this.content.length);
		const newSource = prefix + background("red", this.content) + suffix;
		const lines = newSource.split("\n");
		
		let index = 0;
		for (let i = 0; i < this.position; i++)
			if (this.source[i] === "\n") index++;
		
		const startIndex = Math.max(0, index - 1);

		const excerptLines = lines.slice(
			startIndex,
			Math.min(lines.length, index + 2)
		);

		const maxWidth = String(excerptLines.length + startIndex).length;

		const excerpt = excerptLines
			.map((line, i) => `${String(i + startIndex + 1).padStart(maxWidth)} | ${line}`)
			.join("\n");

		const bar = "=".repeat(40);
		console.log(`\n\n${bar}\n${excerpt}\n${bar}\n${message} (line ${index + 1})\n\n`);
		throw new SyntaxError(message);
		// throw new SyntaxError(message + "\n\n" + excerpt);
	}

	toString() {
		return `(${this.type.toString()}: ${color("blue", this.content)})`;
	}
}

class TokenStream {
	constructor(tokens = []) {
		this.tokens = [...tokens].reverse();
	}

	get length() {
		return this.tokens.length;
	}

	get all() {
		return [...this.tokens].reverse();
	}

	copy() {
		return new TokenStream(this.all);
	}

	prepend(token) {
		this.tokens.push(token);
	}

	has(content, index = 0) {
		if (index >= this.tokens.length)
			return false;
		
		const token = this.tokens[this.tokens.length - index - 1];
		if (typeof content === "string")
			return token.content === content;
		return token.type === content;
	}

	hasAny(...options) {
		let index = 0;
		if (typeof options[options.length - 1] === "number")
			index = options.pop();

		for (let i = 0; i < options.length; i++)
			if (this.has(options[i], index))
				return true;

		return false;
	}

	get(index = 0) {
		return this.getToken(index).content;
	}

	getToken(index = 0) {
		if (index >= this.tokens.length)
			throw new RangeError("Desired index is out of bounds");
		return this.tokens[this.tokens.length - index - 1];
	}

	skip(amount) {
		if (amount > this.tokens.length)
			throw new RangeError("Cannot skip over tokens in an empty stream");
		this.tokens.length -= amount;
	}
	
	skipAll(tok) {
		while (this.has(tok)) this.next();
	}

	remove(content) {
		if (typeof content === "string")
			this.tokens = this.tokens.filter(token => token.content !== content);
		else
			this.tokens = this.tokens.filter(token => token.type !== content);
	}

	nextToken() {
		if (!this.tokens.length)
			throw new RangeError("Cannot advance an empty stream");

		return this.tokens.pop();
	}

	next(expected) {
		if (!this.tokens.length)
			throw new RangeError("Cannot advance an empty stream");
		
		const token = this.tokens.pop();

		if (expected !== undefined) {
			if (typeof expected === "string") {
				if (token.content !== expected)
					token.error(`Unexpected token '${token.content}', expected '${String(expected)}'`);
			} else {
				if (token.type !== expected)
					token.error(`Unexpected token '${token.content}', expected token of type '${String(expected)}'`);
			}
		}

		return token.content;
	}

	optional(content) {
		if (this.has(content)) {
			this.next();
			return true;
		}

		return false;
	}

	until(tok) {
		const result = [];
		while (this.tokens.length && !this.has(tok))
			result.push(this.nextToken());
		return new TokenStream(result);
	}

	endOf(open, close) {
		const result = [];

		this.until(open);
		if (!this.tokens.length)
			throw new RangeError(`The specified boundaries "${open}${close}" don't exist`);
		this.next();

		let depth = 1;
		while (this.tokens.length && depth) {
			if (this.has(open)) depth++;
			if (this.has(close)) depth--;
			result.push(this.nextToken());
		}

		result.pop();
		return new TokenStream(result);
	}

	delimitedList(parseItem, delimiter, interrupt) {
		const results = [];

		while (this.tokens.length) {
			results.push(parseItem(this));

			if (interrupt !== undefined && this.has(interrupt))
				break;

			if (this.tokens.length)
				this.next(delimiter);
		}

		return results;
	}

	toString() {
		return this.all.join(" ");
	}
}

class TokenStreamBuilder {
	constructor(source) {
		this.source = source;
		this.index = 0;
		this.tokens = [];
	}

	get stream() {
		return new TokenStream(this.tokens);
	}

	append(content, type) {
		const position = this.source.indexOf(content, this.index);
		this.index = position + content.length;
		this.tokens.push(new Token(content, type, position, this.source));
	}

	static regex(source, regexes) {
		const builder = new TokenStreamBuilder(source);

		while (source.length) {
			source = source.replace(/^\s*/, "");
			for (let i = 0; i < regexes.length; i++) {
				const [regex, type] = regexes[i];
				if (regex.test(source)) {
					builder.append(source.match(regex)[0], type);
					source = source.replace(regex, "");
					break;
				}
			}
		}

		return builder.stream;
	}
}